resources/scripts/ttx_diff.py

```diff
@@ -519,9 +520,60 @@ def remove_gdef_lig_caret_and_var_store(ttx: etree.ElementTree):
         if subtable is not None:
             gdef.remove(subtable)
 
+# remap the elements, and return a map of new_ix -> prev_ix
+def remap_class_def_ids(class_def: etree.ElementTree) -> Dict[int, int]:
+    current_classes = defaultdict(list)
+    for  glyph in class_def.xpath(".//ClassDef"):
+        cls = glyph.attrib["class"]
+        current_classes[cls].append(glyph.attrib['glyph'])
+
+    # match the sorting used in fonttools:
+    # https://github.com/fonttools/fonttools/blob/8a89f4f81b0068/Lib/fontTools/otlLib/builder.py#L2689
+    new_order = sorted(current_classes.values(), key=lambda s: (-len(s), s))
+    new_order_map = {name: i + 1 for (i, cls) in enumerate(new_order) for name in cls}
+    result = dict()
+    for  glyph in class_def.xpath(".//ClassDef"):
+        cls = glyph.attrib["class"]
+        new = new_order_map.get(glyph.attrib['glyph'])
+        glyph.attrib["class"] = str(new)
+        result[new] = int(cls)
+    return result
+
+
+def reorder_rules(lookup: etree.ElementTree, new_order: Dict[int, int], rule_name: str):
+    orig_order = [el for el in lookup.iterchildren(tag=rule_name)]
+    prev_el = orig_order[0].getprevious()
+    for el in orig_order:
+        lookup.remove(el)
+
+    for ix in range(len(orig_order)):
+        prev_ix = new_order.get(ix, ix)
+        el = orig_order[prev_ix]
+        el.set("index", str(ix))
+        prev_el.addnext(el)
+        prev_el = el
+
+    etree.indent(lookup, level=4)
+
+
+# fontmake and fontc assign glyph classes differently for class-based tables;
+# fontc uses GIDs but fontmake uses glyph names, so we reorder them to be consistent.
+def reorder_gsub_class_based_rules(ttx: etree.ElementTree):
+    gsub = ttx.find("GSUB")
+    if gsub is None:
+        return
+    for lookup in gsub.xpath(".//Lookup"):
+        chain_ctx = lookup.find("ChainContextSubst")
+        if chain_ctx is None or int(chain_ctx.attrib["Format"]) != 2:
+            continue
+        new_class_order = remap_class_def_ids(chain_ctx.find("InputClassDef"))
+        reorder_rules(chain_ctx, new_class_order, "ChainSubClassSet")
```

Comment:
This feels like it wants to be one function; I'm failing to spot why it can't be?

Reply:
I expect there will be other lookups where rules are ordered based on class indices, and this logic will be reused.